lets try to update node-semver

memsharded opened the pull request:
Changelog: Fix: Upgrade node-semver python library.
Docs: omit

Fix https://github.com/conan-io/conan/issues/6057
Fix https://github.com/conan-io/conan/issues/6459

Related:
- https://github.com/conan-io/conan/issues/5366
- https://github.com/conan-io/conan/issues/2172

memsharded committed:
lets try to update node-semver

memsharded commented:
An interesting question @jgsogo is if we should force 0.8.0 to ensure a consistent behavior of version-range resolution in different Conan installations

jgsogo commented:
As it is a `0.x.y` version I think we should force a fixed version of the library. We don't know what can change in the next minor.

---

There other issues about this topic:
 * https://github.com/conan-io/conan/issues/5391
 * https://github.com/conan-io/conan/issues/5366

As you can read I vote to be deterministic even if the [SemVer documentation](https://semver.org/) doesn't consider microversions (or metadata). In fact, they have moved from SHOULD to MUST "ignore metadata" since them. Changing this means that it is no longer SemVer.

Is it breaking to move to something more deterministic? It wasn't because it depends on the order of the libraries retrieved from the server (was/is that order deterministic)?

uilianries commented:
node-semver 0.8.0 changed the aspect when dealing to versions with more then 3-digits: https://github.com/podhmo/python-semver/issues/34

And what surprise, people asked because of Conan ...

The implementation is here: https://github.com/podhmo/python-semver/pull/35

What has changed? Now, comparing 3-digit to 4-digit will return as different versions. Before that, they were considered the same. Example: 1.1.1 == 1.1.1.1

uilianries committed:
#6057 Fix test for version comparing

Since node-semver 0.8.0, versions with different
number of digits are considered not equal. For instance,
1.1.1 == 1.1.1.1 until 0.7.0 was considered Equal, however
this rule changed in 0.8.0 and now is consired False.

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded commented:
> What has changed? Now, comparing 3-digit to 4-digit will return as different versions. Before that, they were considered the same. Example: 1.1.1 == 1.1.1.1

Can we consider a bug that this happened in the past? Can you submit a fix to Conan to see what it means on our side? Thanks!

uilianries committed:
#6057 Force minimal node-semver version

Approving both node-semver 0.6.1 and 0.8.0 will
cause different behaviors when intalling both
versions.

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded committed:
Merge pull request #30 from uilianries/feature/upgrade_node_semver

#6057 Fix test for version comparing

memsharded commented:
Ok, we have a problem. We cannot upgrade to 0.8.0 because it does not support Python 2.7 anymore. Although Python 2.7 support in Conan is deprecated, it is not completely removed yet, and we want to keep it running at the moment. It seems this will need to wait until Conan 2.0?

uilianries commented:
> Can we consider a bug that this happened in the past? Can you submit a fix to Conan to see what it means on our side? Thanks!

Yes, because we were allowing different versions when comparing them. e.g. 1.1.1 == 1.1.1.1 are different versions, but in the past it was true. If we use 0.8.0 we don't need a fix, because it's embedded in nose-semver. However, if we need to keep 0.6.0, we will need to fix it in Conan code.

> Ok, we have a problem. We cannot upgrade to 0.8.0 because it does not support Python 2.7 anymore. Although Python 2.7 support in Conan is deprecated, it is not completely removed yet, and we want to keep it running at the moment. It seems this will need to wait until Conan 2.0?

I see the perfect excuse for deprecating python 2.7 haha, well, let's keep 0.6.1 and fix it in Conan base code.

uilianries committed:
#6057 Compare micro versions

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
#6057 Update micro property for Version

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded committed:
Merge pull request #31 from uilianries/feature/upgrade_node_semver

#6057 Compare micro versions

uilianries commented on conans/requirements.txt:
```diff
@@ -6,7 +6,7 @@ PyYAML>=3.11, <6.0
 fasteners>=0.14.1
 six>=1.10.0,<=1.14.0
-node-semver==0.6.1
+node-semver>=0.6.1, <0.8.0
```
```suggestion
node-semver>=0.6.1, <=0.8.0
```

@memsharded we shouldn't ignore 0.8.0. Both 0.6.1 and 0.8.0 will result in the same behavior.

memsharded replied:
Yes, but we should probably put the limit on 0.8.0. As it is unstable 0.X, future versions can break. Please add <=0.8.0

uilianries replied:
Okay, @memsharded you can merge it directly by this comment

memsharded commented on conans/client/tools/version.py:
```diff
@@ -42,9 +46,14 @@ def build(self):
         if not isinstance(other, Version):
             other = Version(other)
-        return self._semver.compare(other._semver) == 0
+        return (self._semver.compare(other._semver) or self._compare_micro(other)) == 0
```
So the key is that node-semver already had the ``_compare_micro()`` in 0.6.1?

uilianries replied:
No, it has been included in 0.8.0. This is "port" for 0.6.1

memsharded committed:
Update conans/requirements.txt

Co-authored-by: Uilian Ries <uilianries@gmail.com>

jgsogo commented:
This implies that we no longer follow SemVer, with 4 numbers I've read it as "Explicit versioning" and it is not adding one number to the right, but one to left (non-breaking behavior is 1.2.X.Y), are we going to change package_id_modes?

If we are changing this I would suggest considering lexicographic order to build metadata as well, will it fit into _"Explicit versioning"_ or is it another just another different schema?

memsharded commented:
> This implies that we no longer follow SemVer, with 4 numbers I've read it as "Explicit versioning" and it is not adding one number to the right, but one to left (non-breaking behavior is 1.2.X.Y), are we going to change package_id_modes?

Not sure what you mean with this, can you please clarify?

uilianries commented:
I think it's because we have two Version classes, one in model, and one in tools. But this one is about tools, only for comparing project versions.

jgsogo commented:
We are building a dependency manager, managing versions is one of our core features, we should think very carefully about any change here.

 * SemVer states that only MAJOR, MINOR, PATCH and pre-release have to be taken into account. Any other thing MUST not be taken into account. So this will no longer be SemVer.
 * Is there any SemVerXXX specification that includes MICRO versions, or are we on our own if we include it?
 * When I search on the Internet about 4-digit versions, I come across _"Explicit versioning"_ where the 4th number is not a MICRO version (added to the right: x.y.z.MICRO) but a DISRUPTIVE one (added to the left: DISRUPTIVE.x.y.z). Using this specification, packages are API-compatible if they share the first to numbers.

We are a package manager and we can follow one specification or we can create our own. But before breaking the one we are following right now, **we need to know where we are moving to**.

uilianries commented:
@jgsogo Now I see your point. If we include a 4th digit we would be breaking the SemVer specification.

Checking in semver repo, I've found many people asking for 4th digit support, but this won't happen: https://github.com/semver/semver/issues/323#issuecomment-481450266

So considering it, node-semver is broken in 0.8.0

SSE4 commented:
4 digit is something very common, especially in Microsoft/Windows world. it's usually noted as **major.minor.build.private** (or **major.minor.build.revision**), and usually all resources, DLLs, etc. have these 4 version numbers. I think this is even part of PE/COFF specification.

uilianries commented:
@SSE4 Kernel Linux also uses 4 digits, but it's not part of semver specification. The method tools.Version() uses semver under the hood.

The only close option is pre-release: https://semver.org/#spec-item-9

memsharded commented:
> 4 digit is something very common, especially in Microsoft/Windows world. it's usually noted as major.minor.build.private (or major.minor.build.revision), and usually all resources, DLLs, etc. have these 4 version numbers. I think this is even part of PE/COFF specification.

The problem is that it might be a common thing for specifying versions. But I have been having a look too, and nothing out there defines any kind of version-range resolution against 4 digits versions. Some people might extrapolate the 4th version on the right, and make it a micro version, but this is not defined in the semver especification.

SSE4 commented:
indeed, this is not SemVer, and I think these are versioning schemes predating the SemVer. that particular 4-digit scheme appears to be very common, especially for old(er) projects, and such conventions are frequently used by companies in enterprise world. as a package manager serving both OSS and Enterprise cases, we should probably live with it, and support both versioning schemes, or maybe even arbitrary versioning schemes (e.g. if someone is using latin letters, or roman numbers, or cyrillic letters, or just arbitrary names). this may require us to: 1) expose several version classes (e.g. SemVer, FourDigitVer, etc.) 2) allow author to provide/expose his own comparators/operators to handle his versioning gracefully

puetzk commented:
> I think this is even part of PE/COFF specification.

It's not part of PE/COFF, but it **is** part of the standard Win32 version resource, and it's also a required part of doing WinSxS manifest [assembly version](https://docs.microsoft.com/en-us/windows/win32/sbscs/assembly-versions) (where you *must* use 4 digits, if only by appending zeros). It's also CMake's default scheme for `PROJECT_VERSION` (where they call it major.minor.patch.tweak).

memsharded commented:
This is very risky. Taking it for Conan 2.0

memsharded commented:
This was too risky for Conan 1.X. Conan 2.0 in https://github.com/conan-io/conan/pull/10058 is removing the external dependency to node-semver, and implementing our own version and version-range management. Closing this.
